api/deploy.go

```diff
@@ -83,31 +84,41 @@ func newServiceDeployer(api *API, env map[string]string, options ...DeployServic
 	return d
 }
 
-// FromGitURL deploys a service hosted at a Git url.
-func (d *serviceDeployer) FromGitURL(url string) (*service.Service, *importer.ValidationError, error) {
-	defer d.closeStatus()
-	d.sendStatus("Downloading service...", Running)
-	path, err := d.createTempDir()
-	if err != nil {
-		return nil, nil, err
-	}
-	defer os.RemoveAll(path)
-	if err := xgit.Clone(url, path); err != nil {
-		return nil, nil, err
-	}
+// FromURL deploys a service from git or tarball url.
+func (d *serviceDeployer) FromURL(url string) (*service.Service, *importer.ValidationError, error) {
+	if xgit.IsGitURL(url) {
+		defer d.closeStatus()
+		d.sendStatus("Downloading service...", Running)
+		path, err := d.createTempDir()
+		if err != nil {
+			return nil, nil, err
+		}
+		defer os.RemoveAll(path)
+		if err := xgit.Clone(url, path); err != nil {
+			return nil, nil, err
+		}
 
-	// XXX: remove .git folder from repo.
-	// It makes docker build iamge id same between repo clones.
-	if err := os.RemoveAll(filepath.Join(path, ".git")); err != nil {
-		return nil, nil, err
+		// XXX: remove .git folder from repo.
+		// It makes docker build iamge id same between repo clones.
+		if err := os.RemoveAll(filepath.Join(path, ".git")); err != nil {
+			return nil, nil, err
+		}
+
+		d.sendStatus("Service downloaded with success", DonePositive)
+		r, err := xarchive.GzippedTar(path, nil)
+		if err != nil {
+			return nil, nil, err
+		}
+		return d.deploy(r)
 	}
 
-	d.sendStatus("Service downloaded with success", DonePositive)
-	r, err := xarchive.GzippedTar(path, nil)
+	// if not git repo then it should be tarball
+	resp, err := http.Get(url)
 	if err != nil {
 		return nil, nil, err
 	}
-	return d.deploy(r)
+	defer resp.Body.Close()
+	return d.FromGzippedTar(resp.Body)
```

Comment:
nothing blocking but for convenience, we can use _deploy()_ here and move line 90 _(defer d.closeStatus())_ to 89.